src/github/github.installation.service.ts

```diff
@@ -0,0 +1,173 @@
+    const org = await this.allowedOrgRepository.findOne({
+      where: { organizationName, status: OrganizationStatus.ACTIVE },
+    });
+    return !!org;
+  }
+
+  async isUserAllowed(
+    organizationName: string,
+    githubUsername: string,
+  ): Promise<boolean> {
+    try {
+      const org = await this.findByName(organizationName);
+      if (org.status !== OrganizationStatus.ACTIVE) {
+        return false;
+      }
+      if (org.allowAllUsers) {
+        return true;
+      }
+      return org.activeUsers.includes(githubUsername);
+    } catch (error) {
+      this.logger.error(`Error checking user allowance: ${error.message}`);
+      return false;
+    }
+  }
+
+  async addAllowedUser(
+    organizationName: string,
+    githubUsername: string,
+  ): Promise<AllowedOrganization> {
+    const org = await this.findByName(organizationName);
+    if (!org) {
+      throw new NotFoundException(`Organization not found`);
+    }
+    if (org.inactiveUsers.includes(githubUsername)) {
+      org.inactiveUsers = org.inactiveUsers.filter(
+        (username) => username !== githubUsername,
+      );
+    }
+    if (!org.activeUsers.includes(githubUsername)) {
+      org.activeUsers.push(githubUsername);
+      return this.allowedOrgRepository.save(org);
+    }
+    return org;
+  }
+
+  async removeAllowedUser(
+    organizationName: string,
+    githubUsername: string,
+  ): Promise<AllowedOrganization> {
+    const org = await this.findByName(organizationName);
+    if (!org) {
+      throw new NotFoundException(`Organization not found`);
+    }
+    org.activeUsers = org.activeUsers.filter(
+      (username) => username !== githubUsername,
+    );
+    if (!org.inactiveUsers.includes(githubUsername)) {
+      org.inactiveUsers.push(githubUsername);
+    }
+    return this.allowedOrgRepository.save(org);
+  }
+
+  async updateStatus(
+    organizationName: string,
+    newStatus: OrganizationStatus,
+  ): Promise<AllowedOrganization> {
+    const org = await this.findByName(organizationName);
+    if (!org) {
+      throw new NotFoundException(`Organization not found`);
+    }
+    org.status = newStatus;
+    return this.allowedOrgRepository.save(org);
+  }
+
+  async uninstall(id: number) {
+    const now = Math.floor(Date.now() / 1000);
+
+    const pemFilename = Env.GITHUB.APP_PEM;
+    const pemFilePath = path.join(__dirname, '../../../', pemFilename);
+    const privateKey = fs.readFileSync(pemFilePath, 'utf8');
+
+    const payload = {
+      iat: now,
+      exp: now + 600,
+      iss: Env.GITHUB.APP_ID,
+      installation_id: id,
+    };
+    const token = jwt.sign(payload, privateKey, { algorithm: 'RS256' });
+    const octokit = new Octokit({
+      auth: token,
+    });
+
+    await octokit.request('DELETE /app/installations/{installation_id}', {
+      installation_id: id,
+      headers: {
+        'X-GitHub-Api-Version': '2022-11-28',
+      },
+    });
```

Comment:
PullProofAI

There is a potential performance issue in the `uninstall` method. The synchronous file reading operation using `fs.readFileSync` can block the event loop, leading to poor performance, especially in high-traffic scenarios. It's recommended to use asynchronous file reading to prevent blocking the event loop.

Suggested Improvement:
```typescript
144:   async uninstall(id: number) {
145:     const now = Math.floor(Date.now() / 1000);
146: 
147:     const pemFilename = Env.GITHUB.APP_PEM;
148:     const pemFilePath = path.join(__dirname, '../../../', pemFilename);
149:     
150:     // Asynchronously read the private key file
151:     fs.readFile(pemFilePath, 'utf8', (err, privateKey) => {
152:       if (err) {
153:         // Handle error appropriately
154:         console.error(err);
155:         return;
156:       }
157:       
158:       const payload = {
159:         iat: now,
160:         exp: now + 600,
161:         iss: Env.GITHUB.APP_ID,
162:         installation_id: id,
163:       };
164:       const token = jwt.sign(payload, privateKey, { algorithm: 'RS256' });
165:       const octokit = new Octokit({
166:         auth: token,
167:       });
168:       
169:       octokit.request('DELETE /app/installations/{installation_id}', {
170:         installation_id: id,
171:         headers: {
172:           'X-GitHub-Api-Version': '2022-11-28',
173:         },
174:       }).then(() => {
175:         // Handle successful uninstallation
176:       }).catch((error) => {
177:         // Handle uninstallation failure
178:         console.error(error);
179:       });
180:     });
181:   }
```

This improvement ensures that the file reading operation does not block the event loop, enhancing the performance of the application during file operations.


<!-- This is an auto-generated comment by PullProofAI -->